apps/application-system/api/src/app/modules/application/tools/application.serializer.ts

```diff
@@ -25,22 +27,39 @@ export class ApplicationSerializer
     context: ExecutionContext,
     next: CallHandler,
   ): Observable<Promise<unknown>> {
-    let nationalId: string
-    try {
-      nationalId = getNationalIdFromToken(context)
-    } catch (e) {
-      throw new UnauthorizedException('You are not authenticated')
-    }
+    const user = getCurrentUser(context)
+
     return next.handle().pipe(
       map(async (res: Application | Array<Application>) => {
         const isArray = Array.isArray(res)
-        return isArray
-          ? Promise.all(
-              (res as Application[]).map((item) =>
-                this.serialize(item, nationalId),
-              ),
-            )
-          : this.serialize(res as Application, nationalId)
+
+        if (isArray) {
+          return Promise.all(
+            (res as Application[]).map((item) => {
+              if (
+                item.applicant !== user.nationalId &&
+                !item.assignees.includes(user.nationalId)
+              ) {
+                throw new NotFoundException(
+                  'No application found for this user',
+                )
+              }
+
+              return this.serialize(item, user.nationalId)
+            }),
+          )
+        }
+
+        const application = res as Application
+
+        if (
+          application.applicant !== user.nationalId &&
+          !application.assignees.includes(user.nationalId)
+        ) {
+          throw new NotFoundException('No application found for this user')
+        }
```

Comment:
This check happens to late. It needs to be at the top of the controller handlers instead, so we block access to any update logic.

Maybe we should update `ApplicationController#findOne`, remove `ApplicationByIdPipe` and instead call a helper that fetches and checks access: `this._findOneByIdAndNationalId(id, user.nationalId)`.

Reply:
I did that first. But didn’t know how to pass the nationalID down to the pipeTransform so change to this. But sounds good. I will do that 👍

Reply:
I created a small service for it that contains the function. Wasn't sure if it should live at the root of the controller or into it's own file